broadcast/src/buffered/engine.rs

```diff
@@ -261,13 +326,37 @@ impl<
     /// Returns `true` if the message was inserted, `false` if it was already present.
     /// Updates the deque, item count, and message cache, potentially evicting an old message.
     fn insert_message(&mut self, peer: P, msg: M) -> bool {
+        // Get the identity and digest of the message
+        let identity = msg.identity();
         let digest = msg.digest();
 
         // Send the message to the waiters, if any, ignoring errors (as the receiver may have dropped)
-        if let Some(responders) = self.waiters.remove(&digest) {
-            for responder in responders {
+        if let Some(mut waiters) = self.waiters.remove(&identity) {
+            let mut i = 0;
+            while i < waiters.len() {
+                // Get the sender and digest filters
+                let (sender_filter, digest_filter, _) = &waiters[i];
+
+                // Keep the waiter if either filter does not match.
+                if sender_filter.as_ref().is_some_and(|s| s != &peer)
+                    || digest_filter.is_some_and(|d| d != digest)
+                {
+                    i += 1;
+                    continue;
+                }
+
+                // Filters match → fulfil the subscription and drop the entry.
+                //
+                // The index `i` is intentionally not incremented here to check
+                // the element that was swapped into position `i`.
+                let (_, _, responder) = waiters.swap_remove(i);
                 self.respond_subscribe(responder, msg.clone());
             }
+
+            // Re-insert if any waiters remain for this identity.
+            if !waiters.is_empty() {
+                self.waiters.insert(identity, waiters);
+            }
```

Comment:
Style: consider using the `Entry` API in rust. In combination with my suggestion about `retain_mut`, it might look like this:

```
if let Entry::Occupied(entry) = self.waiters.entry(&identity) {
  let waiters = entry.get_mut();
  waiters.retain_mut(|(s, d, responder)| {
    let not_relevant = s.as_ref().is_some_and(|s| s != &peer)
      || d.is_some_and(|d| d != digest);
    if not_relevant {
      true
    }
    self.respond_subscribe(responder, msg.clone());
    false
  });
  if waiters.is_empty() {
     entry.remove_entry();
  }
}
```

Reply:
`retain_mut` doesn't provide ownership of `responder`, unfortunately. This prevents us from calling `respond_subscribe` and won't work directly (unless I'm missing something).

Reply:
It also doesn't return anything for us to use (like the emitted items or something).

Reply:
That being said, using `entry` here is definitely an improvement and I'll do that.

Reply:
Ah shoot, that's also tough because `respons_subscribe` requires a mutable reference to `self` (just like the entry).